### ISSUE TITLE
When removing ratelimit-server from existing clusters we get not in-sync with the EDS API alerts

### ISSUE THREAD

**Issue by DreyfussDaena** [solo-io/gloo#10267]
### Gloo Edge Product

Enterprise

### Gloo Edge Version

gloo-ee-1.17.x

### Kubernetes Version

v1.28.6

### Describe the bug

Hi Solo team, 

As part of an initiative to cut costs on existing clusters, 
We have tried removing the ratelimit server which is currently installed.
When we do this, we get the following alerts: 
One or more envoy instances are not in-sync with the EDS API for the last 5 minutes.
We get these alerts even though we removed an upstream that no one is referring to.

Please assist

### Expected Behavior

We should be able to remove the ratelimit server without getting any sync alerts.

### Steps to reproduce the bug

At first, we removed its reference from gloo settings resource, the ratelimitServer section:
 
ratelimitServer:
  rateLimitBeforeAuth: false
  ratelimitServerRef:
    name: ugw-ratelimit-server
    namespace: ratelimitsrv
 
This worked without any issues.
 
Then we try to delete the upstream resource ugw-ratelimit-server.
 
This causes alerts: One or more envoy instances are not in-sync with the EDS API for the last 5 minutes
 
We get these alerts even though we removed an upstream that no one is referring to.
 
We tried restating the gloo pods but still get the alerts.

**Comment by DuncanDoyle** [comment]
@DreyfussDaena Where do you see these alerts? In the logs of the Gloo controller pod? I'm trying to reproduce this, but haven't been successful so far. 

Do you install the platform using Helm? When using Helm, there is a configuration option to disable the rate-limit server using Helm values:

```
global:
  extensions:
    # Rate-Limit Configuration
    rateLimit:
      enabled: false
```

Could you try disabling the rate-limit server using that approach?

**Comment by DreyfussDaena** [comment]
Hey @DuncanDoyle 
We aren't using the gloo ratelimit configuration.
we are using the gloo ratelimit-server settings as we are bringing our own ratelimit server.

**Comment by AkshayAdsul** [comment]
I am able to reproduce this issue.
It appears we need atleast one `Endpoints` object. Else the ClusterLoadAssignment("type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment")  values goes to 0.
And we get the below error

```
Gloo has detected that the data plane is out of sync. The following types of resources have not been accepted: [{resource="type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment"}]. Gloo will not be able to process any other configuration updates until these errors are resolved.
```

1. For any standard GE installation( I tried in 1.17.3) in your settings object (`kubectl edit settings.gloo.solo.io -o yaml -n gloo-system`) have `disableKubernetesDestinations: true` and also have the Upstream discovery service options disabled.
```
discovery:
    udsOptions:
      enabled: false
```
2. Make sure you do not have any discovered upstreams .
You can use below command to be sure
``` 
kubectl get upstream -n gloo-system -l discovered_by=kubernetesplugin  --no-headers=true -o json | jq -r '.items[] | select(.metadata.name)| .metadata.name' | xargs  kubectl delete -n gloo-system upstream
```
3.  You could decide to use only [static upstreams](https://docs.solo.io/gloo-edge/latest/guides/traffic_management/destination_types/static_upstream/).

4. Although this should be valid when you do a glooctl check. You get a data plane is out of sync error.

5. You can port-forward the gloo controller pod and grab the metrics and check the ClusterLoadAssignment value.
It will be 0.
 a. kubectl port-forward -n gloo-system deploy/gloo 9091:9091 &
 b. curl -s http://localhost:9091/metrics > gloometrics
 c. cat gloometrics | grep ClusterLoadAssignment
glooe_solo_io_xds_insync{resource="type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment"} 0

6. This causes not in-sync issues which can be confirmed with `glooctl check`. 

7. Create at least one upstream which uses EDS and you will notice that it starts working.
kubectl apply below dummy upstream

```
apiVersion: gloo.solo.io/v1
kind: Upstream
metadata:
  name: dummy
  namespace: gloo-system
spec:
  kube:
    selector:
      gloo: gloo
    serviceName: gloo
    serviceNamespace: gloo-system
    servicePort: 443
```
8. Creating this dummy upstream can be a workaround for the current problem. Need to discuss more if there is a cleaner way to address this.

**Comment by DreyfussDaena** [comment]
@AkshayAdsul 
Can I remove all the related services (+ pod) and leave the upsteam in the system (the ratelimitsrv upstream)?
Or will that cause other issues?

Just thinking of possible workarounds for the time being.
Thanks.

**Comment by AkshayAdsul** [comment]
@DreyfussDaena Yes you can remove all the related services (+ pod) and leave the upstream in the system (the ratelimitsrv upstream). No harm in doing it, you won't get the sync alert  and will help your main task of cost cutting. Thanks.

**Comment by AkshayAdsul** [comment]
If possible the customer has also requested a backport to 1.17 for this. Thanks.

**Comment by DreyfussDaena** [comment]
As suggested, 
We removed all related services an left the uppsteam.
We are getting the following error in the upstream:

apiVersion: gloo.solo.io/v1
kind: Upstream
metadata:
  name: ugw-ratelimit-server
  namespace: ratelimitsrv
spec:
  kube:
    serviceName: ugw-ratelimit-server-service
    serviceNamespace: ratelimitsrv
    servicePort: 18081
    serviceSpec:
      grpc: {}
  sslConfig:
    secretRef:
      name: ugw-ratelimit-server-mtls-certs
      namespace: ratelimitsrv
status:
  statuses:
    apigw-gloo-ee:
      reason: "warning: \n  1 error occurred:\n\t* Upstream name:\"ugw-ratelimit-server\"
        namespace:\"ratelimitsrv\" references the service \"ugw-ratelimit-server-service\"
        which does not exist in namespace \"ratelimitsrv\"\n\n"
      reportedBy: gloo
      state: Warning
      
This doesn't look healthy..
Please advise.

Thanks,
Daena

**Comment by AkshayAdsul** [comment]
I will take a look again when I am in front of my laptop tomorrow. But can you please  try creating the dummy upstream workaround suggested above ?
It’s in the step 7 of the Steps to reproduce portion  of my comment  above.

> Potential Workaround:
Create at least one upstream which uses EDS and you will notice that it starts working.
kubectl apply below dummy upstream

**Comment by DreyfussDaena** [comment]
Hey @AkshayAdsul 
We'd prefer not adding dummy resources...
We removed all the connections and left the existing upstream - effectively creating a dummy upstream and it seems to still fail.
I think it would be better to find a proper solution to this issue.

Thanks, 
Daena

**Comment by AkshayAdsul** [comment]
Hi @DreyfussDaena 
So it still needs to have at least 1 valid upstream . 
When you deleted the services (+ pod),  the ratelimitsrv upstream became invalid. Hence you get that error `reason: "warning: \n 1 error occurred:\n\t* Upstream name:"ugw-ratelimit-server"
namespace:"ratelimitsrv" references the service "ugw-ratelimit-server-service"
which does not exist in namespace "ratelimitsrv"\n\n"`.
So you can delete that ratelimitsrv upstream.
And create a dummy upstream like below 

```
kubectl apply -f - <<EOF
apiVersion: gateway.solo.io/v1
apiVersion: gloo.solo.io/v1
kind: Upstream
metadata:
  name: dummy
  namespace: gloo-system
spec:
  kube:
    selector:
      gloo: gloo
    serviceName: gloo
    serviceNamespace: gloo-system
    servicePort: 443
EOF
```


For a proper solution on the code I will let @nfuden suggest a fix.
